gemini/use-cases/retrieval-augmented-generation/multimodal_rag_langchain.ipynb

```diff
@@ -0,0 +1,1026 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "colab": {
+     "base_uri": "https://localhost:8080/"
+    },
+    "executionInfo": {
+     "elapsed": 1757,
+     "status": "ok",
+     "timestamp": 1707913903524,
+     "user": {
+      "displayName": "",
+      "userId": ""
+     },
+     "user_tz": -330
+    },
+    "id": "KwbL89zcY39N",
+    "outputId": "baa3a478-4d55-4b9c-e02f-6816a1d589b1",
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "# Download documents and images used in this notebook\n",
+    "!gsutil -m rsync -r gs://github-repo/rag/intro_multimodal_rag/ .\n",
+    "print(\"Download completed\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "Ps1G-cCfpibN"
+   },
+   "source": [
+    "## Partition PDF tables, text, and images"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "jqLsy3iZ5t-R"
+   },
+   "source": [
+    "### The data\n",
+    "\n",
+    "The source data that you will use in this notebook is a modified version of [Google-10K](https://abc.xyz/assets/investor/static/pdf/20220202_alphabet_10K.pdf) which provides a comprehensive overview of the company's financial performance, business operations, management, and risk factors. As the original document is rather large, you will be using [a modified version with only 14 pages](https://storage.googleapis.com/github-repo/rag/multimodal_rag_langchain/google-10k-sample-14pages.pdf) instead. Although it's truncated, the sample document still contains text along with images such as tables, charts, and graphs."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "pdf_folder_path = \"/content/data/\" if \"google.colab\" in sys.modules else \"data/\"\n",
+    "pdf_file_name = \"google-10k-sample-14pages.pdf\"\n",
+    "\n",
+    "# Extract images, tables, and chunk text from a PDF file.\n",
+    "raw_pdf_elements = partition_pdf(\n",
+    "    filename=pdf_file_name,\n",
+    "    extract_images_in_pdf=False,\n",
+    "    infer_table_structure=True,\n",
+    "    chunking_strategy=\"by_title\",\n",
+    "    max_characters=4000,\n",
+    "    new_after_n_chars=3800,\n",
+    "    combine_text_under_n_chars=2000,\n",
+    "    image_output_dir_path=pdf_folder_path,\n",
+    ")\n",
+    "\n",
+    "# Categorize extracted elements from a PDF into tables and texts.\n",
+    "tables = []\n",
+    "texts = []\n",
+    "for element in raw_pdf_elements:\n",
+    "    if \"unstructured.documents.elements.Table\" in str(type(element)):\n",
+    "        tables.append(str(element))\n",
+    "    elif \"unstructured.documents.elements.CompositeElement\" in str(type(element)):\n",
+    "        texts.append(str(element))\n",
+    "\n",
+    "# Optional: Enforce a specific token size for texts\n",
+    "text_splitter = CharacterTextSplitter.from_tiktoken_encoder(\n",
```

Comment:
I got the following error here:

```
ImportError: Could not import tiktoken python package. This is needed in order to calculate max_tokens_for_prompt. Please install it with `pip install tiktoken`.
```

So you can add that to your `pip install` line at the top.

Reply:
Good catch, that got missed.